### PR TITLE
Cluster resource tagging

### PR DESCRIPTION
**Pull request by kimorris27**
Edit: I've implemented my proposed solution and pushed it to this PR branch.

---

I'm opening this draft PR because I'm close to being done with policy-based cluster resource tagging, but I've run into a problem and I'd appreciate feedback/suggestions.

### Problem 

I've tested this code out, and install-time tagging looks perfect. The policy gets assigned to the cluster RG before any resources are added, and so each time a new resource is created, it gets the proper tags via the policy's `modify` effect.

But if a customer updates their set of tags, the policy doesn't automatically update the resource tags accordingly. The RP needs to trigger a [remediation task](https://learn.microsoft.com/en-us/azure/governance/policy/how-to/remediate-resources?tabs=azure-portal) to tell Azure Policy that it wants the policy to update resource tags as needed.

The problem I've found while testing it out is that updating the policy assignment has an undefined lag within the Azure cloud, meaning that when we trigger the remediation task immediately after updating the policy assignment, the remediation won't necessarily be acting on the up-to-date tag set. Azure Policy doesn't have any "create and wait for the changes to propagate everywhere" types of operations - you just have to deal with the lag.

I've done a few test updates to watch what happens, and the remediation task only adds the new tags to a subset of the resources. This means that for customers to ensure their resource tags are up-to-date after adding new ones, they would basically need to run `az aro update` repeatedly until the policy assignment has propagated out correctly and the remediation tasks cover all of the resources.

TLDR: Azure Policy is somewhat laggy, so remediating cluster resource tags via policy remediation tasks post-install takes an undefined amount of time and repeated retries (at least with this code).

### Ideas I've had that I don't think will work

- Before triggering a remediation task, poll the policy assignment until it's up-to-date. - :x: - The API call to create/update a policy assignment is not a long-running operation, and the lag we are dealing with is internal to Azure Policy and isn't "pollable". See https://learn.microsoft.com/en-us/azure/governance/policy/how-to/get-compliance-data#evaluation-triggers for info about policy assignment timing (it says it takes about 5 minutes, but the timing is not formally defined).
- Run remediation tasks repeatedly until all resources in the RG have been updated. - :x: - One remediation task takes about 15 minutes, so a single `az aro update` command could take 30 minutes to an hour (maybe even longer?). Horrible for UX and the RP's performance.

### An idea I've had that might work

I think this will work, but I'm open to other suggestions.

Rather than remediating tags post-install time with policy remediation tasks, we can just remediate them "manually" by iterating over all resources in the resource group and adding any tags that the customer has newly added. In this way:
- Existing resources' tags are remediated quickly.
- The policy assignment still gets updated such that any new resources created later on following the cluster update will include the newly-added tags.

The Azure Go SDK has functionality that will make it easy to iterate over the resources and update their tags in a generic way: https://pkg.go.dev/github.com/azure/azure-sdk-for-go@v63.1.0+incompatible/services/resources/mgmt/2019-07-01/features#ResourcesClient.CreateOrUpdateByID.

Please let me know what you think!

### CODE REVIEW

File: pkg/cluster/deploystorage.go [hunk 1]
```diff
@@ -120,6 +154,112 @@ func (m *manager) ensureResourceGroup(ctx context.Context) (err error) {
...
+	resources, err := m.resources.ListByResourceGroup(ctx, resourceGroup, "", "", nil)
+
+	if err != nil {
+		return err
+	}
+
+	for _, r := range resources {
+		// Do not remove any tags, but add new tags and update values
+		// for existing tags.
+		for k, v := range m.doc.OpenShiftCluster.Properties.ResourceTags {
```
Comment (cadenmarchese) [inline pkg/cluster/deploystorage.go:266]: Once concern I had thinking about this approach is that we're putting all of the tag remediation work on the RP rather than Azure. So if the customer has 10,000 disks or something, the RP has a lot of work to do.

That being said, it looks like you're waiting for completion before proceeding in `UpdateByIDAndWait` which I think is the right move assuming we go this direction.

File: pkg/cluster/deploystorage.go [hunk 2]
```diff
@@ -265,3 +405,85 @@ func generateInfraID(base string, maxLen int, randomLen int) string {
...
 	return fmt.Sprintf("%s-%s", base, utilrand.String(randomLen))
 }
+
+func resourceTaggingPolicyDefinition(displayName string) mgmtpolicy.Definition {
+	dp := &mgmtpolicy.DefinitionProperties{}
+
+	dp.PolicyType = mgmtpolicy.Custom
+	dp.Mode = to.StringPtr("Indexed")
+	dp.DisplayName = &displayName
+	dp.Description = to.StringPtr("This policy definition was created via the Azure Red Hat OpenShift RP and is used to tag the resources in your ARO cluster's RP-managed resource group. *Do not delete or modify this policy definition.* The ARO RP will delete this policy definition when/if it deletes the associated cluster.")
```
Comment (cadenmarchese) [inline pkg/cluster/deploystorage.go:444]: Do we need to make any adjustments to deny assignment or as part of this? https://github.com/Azure/ARO-RP/blob/master/pkg/cluster/deploystorage_resources.go#L45

**Review by cadenmarchese — COMMENTED** [review]
On first pass this looks good to me! Some other misc thoughts so far:

- Do we want to set an upper limit on how many resources we tag at a time to avoid the RP having to loop over thousands of potential resources
- Do we need the new API in a `preview` state so that we can deprecate this if/when openshift puts similar functionality in the platform
- Any changes to the deny assignment needed with this?
- Tag deletion (say a customer has multiple and they only want to remove 1 tag) needs to be clearly documented

File: pkg/cluster/deploystorage.go (hunk 1 above)
Comment (kimorris27) [reply pkg/cluster/deploystorage.go:266]: One thought I had is that if the # resources in the cluster RG is greater than a certain amount, we could have `az aro update` error out and ask the customer to run a policy remediation task themselves. This would offload the work to Azure in those extreme cases.

File: pkg/cluster/deploystorage.go (hunk 2 above)
Comment (kimorris27) [reply pkg/cluster/deploystorage.go:444]: Policy definitions can only be created at subscription scope, so we unfortunately can't leverage the deny assignment here.
